File: test/root.test.ts
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import {
  CANDIDATE_ROOTS,
  DEFAULT_ROOT,
  resolveRootDir,
  rootSegment,
} from "../src/root";

const tempDirs: string[] = [];

function makeRepoRoot(): string {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), "plan-sync-root-"));
  tempDirs.push(dir);
  return dir;
}

function initRoot(repoRoot: string, name: string): void {
  const dir = path.join(repoRoot, name);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(path.join(dir, ".sync-config.json"), "{}");
}

afterEach(() => {
  while (tempDirs.length > 0) {
    const dir = tempDirs.pop();
    if (dir) {
      fs.rmSync(dir, { recursive: true, force: true });
    }
  }
});

describe("resolveRootDir", () => {
  it("falls back to the default root when nothing is initialized", () => {
    expect(resolveRootDir(makeRepoRoot())).toBe(DEFAULT_ROOT);
  });

  it("auto-detects the sole initialized candidate root", () => {
    const repoRoot = makeRepoRoot();
    initRoot(repoRoot, ".omx");
    expect(resolveRootDir(repoRoot)).toBe(".omx");
  });

  it("ignores a candidate directory without a .sync-config.json", () => {
    const repoRoot = makeRepoRoot();
    fs.mkdirSync(path.join(repoRoot, ".omx"), { recursive: true });
    expect(resolveRootDir(repoRoot)).toBe(DEFAULT_ROOT);
  });

  it("falls back to the default root when detection is ambiguous", () => {
    const repoRoot = makeRepoRoot();
    initRoot(repoRoot, ".omx");
    initRoot(repoRoot, ".adlc");
    expect(resolveRootDir(repoRoot)).toBe(DEFAULT_ROOT);
  });

  it("prefers an explicit --root over auto-detection", () => {
    const repoRoot = makeRepoRoot();
    initRoot(repoRoot, ".omx");
    expect(resolveRootDir(repoRoot, "  .adlc  ")).toBe(".adlc");
  });

  it.each([" ", ".", "..", "/etc", "a/b", "a\\b", "../escape"])(
    "rejects the unsafe --root value '%s'",
    (value) => {
      expect(() => resolveRootDir(makeRepoRoot(), value)).toThrow();
    },
  );

  // Regression: '...' is not '.' or '..', contains no separators, and is not
  // absolute — so it used to validate — but rootSegment() strips one leading
  // dot and returned '..', escaping a directory level wherever the segment is
  // joined onto a path (e.g. under PLAN_SYNC_STATE_DIR) and yielding an
  // invalid git refname.
  it("rejects --root '...' instead of silently turning it into '..'", () => {
    const repoRoot = makeRepoRoot();
    expect(() => resolveRootDir(repoRoot, "...")).toThrow(
      /strips to the unsafe namespace segment '\.\.'/,
    );
    expect(() => resolveRootDir(repoRoot, "  ...  ")).toThrow(
      /strips to the unsafe namespace segment '\.\.'/,
    );
  });
});

describe("rootSegment", () => {
  it("strips exactly one leading dot from each candidate root", () => {
    expect(CANDIDATE_ROOTS.map(rootSegment)).toEqual(["omc", "omx", "adlc"]);
  });

  it("leaves a dotless root unchanged", () => {
    expect(rootSegment("omc")).toBe("omc");
  });

  it.each(["...", "..", ".", "", ".a/b", ".a\\b"])(
    "throws rather than returning the unsafe stripped segment for '%s'",
    (value) => {
      expect(() => rootSegment(value)).toThrow(
        /strips to the unsafe namespace segment/,
      );
    },
  );
});

File: go/internal/root/root_test.go
package root

import (
	"os"
	"path/filepath"
	"testing"
)

// initRoot creates repoRoot/<name>/.sync-config.json, marking that candidate
// as an initialized root for auto-detection.
func initRoot(t *testing.T, repoRoot, name string) {
	t.Helper()
	dir := filepath.Join(repoRoot, name)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		t.Fatalf("mkdir %s: %v", dir, err)
	}
	if err := os.WriteFile(filepath.Join(dir, syncConfigFile), []byte("{}"), 0o644); err != nil {
		t.Fatalf("write sync config in %s: %v", dir, err)
	}
}

func TestResolveRootDirDefaultsWhenNothingInitialized(t *testing.T) {
	got, err := ResolveRootDir(t.TempDir(), "")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != DefaultRoot {
		t.Fatalf("got %q, want %q", got, DefaultRoot)
	}
}

func TestResolveRootDirAutoDetectsSoleInitializedRoot(t *testing.T) {
	repoRoot := t.TempDir()
	initRoot(t, repoRoot, ".omx")

	got, err := ResolveRootDir(repoRoot, "")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != ".omx" {
		t.Fatalf("got %q, want %q", got, ".omx")
	}
}

func TestResolveRootDirIgnoresCandidateWithoutSyncConfig(t *testing.T) {
	repoRoot := t.TempDir()
	if err := os.MkdirAll(filepath.Join(repoRoot, ".omx"), 0o755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}

	got, err := ResolveRootDir(repoRoot, "")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != DefaultRoot {
		t.Fatalf("got %q, want %q", got, DefaultRoot)
	}
}

func TestResolveRootDirFallsBackWhenAmbiguous(t *testing.T) {
	repoRoot := t.TempDir()
	initRoot(t, repoRoot, ".omx")
	initRoot(t, repoRoot, ".adlc")

	got, err := ResolveRootDir(repoRoot, "")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != DefaultRoot {
		t.Fatalf("got %q, want %q", got, DefaultRoot)
	}
}

func TestResolveRootDirPrefersExplicitRootOverDetection(t *testing.T) {
	repoRoot := t.TempDir()
	initRoot(t, repoRoot, ".omx")

	got, err := ResolveRootDir(repoRoot, "  .adlc  ")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != ".adlc" {
		t.Fatalf("got %q, want %q", got, ".adlc")
	}
}

func TestValidateRootRejectsUnsafeValues(t *testing.T) {
	for _, value := range []string{
		"",
		"   ",
		".",
		"..",
		"/etc",
		"a/b",
		`a\b`,
		"../escape",
	} {
		if got, err := ValidateRoot(value); err == nil {
			t.Errorf("ValidateRoot(%q) = %q, want error", value, got)
		}
	}
}

// TestValidateRootRejectsDotStripEscape is the core regression: "..." passes
// every naive check (it is not "." or "..", has no separators, is not
// absolute) yet RootSegment strips one leading dot and yields "..", escaping
// a directory level wherever the segment is joined onto a path and producing
// an invalid git refname.
func TestValidateRootRejectsDotStripEscape(t *testing.T) {
	repoRoot := t.TempDir()

	for _, value := range []string{"...", "  ...  "} {
		got, err := ResolveRootDir(repoRoot, value)
		if err == nil {
			t.Errorf("ResolveRootDir(--root %q) = %q, want error", value, got)
		}
		if got, err := ValidateRoot(value); err == nil {
			t.Errorf("ValidateRoot(%q) = %q, want error", value, got)
		}
	}
}

func TestRootSegmentStripsOneLeadingDot(t *testing.T) {
	cases := map[string]string{
		".omc":  "omc",
		".omx":  "omx",
		".adlc": "adlc",
		"omc":   "omc",
	}
	for in, want := range cases {
		got, err := RootSegment(in)
		if err != nil {
			t.Fatalf("RootSegment(%q): unexpected error: %v", in, err)
		}
		if got != want {
			t.Errorf("RootSegment(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestRootSegmentRejectsUnsafeStrippedResult(t *testing.T) {
	for _, value := range []string{"...", "..", ".", "", ".a/b"} {
		got, err := RootSegment(value)
		if err == nil {
			t.Errorf("RootSegment(%q) = %q, want error", value, got)
		}
		if got != "" {
			t.Errorf("RootSegment(%q) returned %q alongside its error", value, got)
		}
	}
}
